### DISCUSSION TITLE
In GCP Dataform, how to see Javascript execution log ?

### DISCUSSION THREAD

**Discussion by Riduidel** [dataform-co/dataform#1543] [Q&A] [answered]
Suppose I write a JS file for some complex processing (which is, as far as I understand, the precise use case). Where can I find the results of a `console.log` call? I've searched in Google Cloud Logs, but found nothing. Is it just eaten by the system?

**Comment by BenBirt** [comment, marked as answer]
Console output is indeed not available.

However, just to avoid confusion: Dataform JS code is not executed at workflow execution time - it is executed at compile time, i.e. turning the JS into pure SQL (which can then be executed at some later date). So it's not as though you could e.g. have a console.log outputting some text as a result of running some SQL.

**Reply by Riduidel** [reply]
Oh ok ... this limits the features that Javascript can provide to some query templating mechanism, if I understand correctly ...

**Reply by BenBirt** [reply]
That's more or less correct, yes.

JavaScript can be used to dynamically generate the SQL (templating included) and also dynamically generate the workflow graph.

If something dynamic needs to happen at workflow execution time, we would recommend use of BQ scripting.

**Reply by Riduidel** [reply]
I'll have another question about that ;-)

**Reply by Riduidel** [reply]
You talk about BQ scripting, but is it available in GCP Dataform?

**Reply by BenBirt** [reply]
All BQ SQL is available in Dataform.
